rfc/datawherehouse.md

```diff
@@ -0,0 +1,198 @@
+
+> 3. We don't have bucket events in Cloudflare, so need the client to tell us when it has uploaded something to the provided write target.
+
+Actually we can even extend on this point by saying that today we have no verifiability on data being sent by the user, as well as received by the service. By having the client to sign that bytes were sent, and the service to check and also sign that is true will allow us to achieve that. Moreover, we also open the doors in this interaction for a challengen/proof of delivery.
+
+## High level flow of proposed solution
+
+* Simulate bucket event by having client submit `store/deliver` invocation.
+    * Confirms successful transfer of data.
+    * Effect linked by `store/add` receipt.
+    * Can be batched with other invocations like `filecoin/offer`
+* Handler writes to location claims with encoded information
+
+The following diagram presents the described flow, and is illustrated with the following steps:
+
+1. Client requests to store some bytes with the storefront service
+2. Service issues a receipt stating the URI of a write target for the client to write those bytes
+3. Client writes the bytes into the write target
+4. Client notifies the service that the bytes were written to the provided write target
+5. Service verifies that the bytes are stored in the provided write target and writes a claim about it.
+6. Service issues a receipt stating bytes are being stored by the service.
+
+![datawherehouse1](./datawherehouse/datawherehouse-1.svg)
+
+On the other side of things, the next diagram presents the flow when a client wants to read some data, and is illustrated with the following steps:
+
+1. Client request to read some bytes by the CID of the data
+2. Service discovers where the requested bytes are stored relying on content claims service and the materialized claims from `datawherehouse`
+3. Service serves data stored on the discovered location.
+
+![datawherehouse2](./datawherehouse/datawherehouse-2.svg)
+
+## Location claims
+
+Content claims service is currently deployed implementing the [content claims spec](https://github.com/web3-storage/specs/pull/86). Among other claims, it provides [Location Claims](https://hackmd.io/IiKMDqoaSM61TjybSxwHog?view#Location-Claims) which MAY be used to claim that the bytes that hash to a given CID are available in a given URL.
+
+In w3s the service is responsible to deciding the write target, therefore service SHOULD be responsible for claiming the location of the bytes.
+
+While thinking about using location claims to record where bytes are stored by the service, there are a few characteristics we want to have:
+- location claim MUST resolve to a public and fetchable URLs
+- location in location claim SHOULD (ideally) not change recurrently given it MAY impact negatively the reputation of a party
```

Comment:
I really think we should let user choose how long the location claim should be valid when they request content to be published. If they're chosen TTL is to long we can refuse, if it is too short we can also refuse. But broadly speaking it should be a content owners call and not ours.

Again not suggesting to implement all this, but framing it through this terms seem like a good idea